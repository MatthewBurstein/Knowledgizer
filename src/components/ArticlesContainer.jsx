import React, { Component } from 'react';
import Article from './Article';
import Controls from './Controls';
import { delayAndRepeat } from './helpers/ArticlesContainerHelpers.js';

class ArticlesContainer extends Component {
  constructor() {
    super();

    this.shouldPrint = { print : true }
    this.articlesInQueue = [];
    this.state = {
      articles: []
    };
  };

  componentDidMount() {
    this.printArticles();
  };

  printArticles = () => {
    this.callApi()
      .then(res => {
        this.articlesInQueue = this.articlesInQueue.concat(res.articles);
        delayAndRepeat(this.addArticleToState, this.articlesInQueue, this.shouldPrint, 1000);
      })
      .catch(err => console.log('error in callApi is ', err));
  }

  addArticleToState = article => {
    this.setState(prevState => ({ articles: [...prevState.articles, article] }));
  };

  async callApi() {
    const response = await fetch('/api');
    const body = await response.json();
    if (response.status !== 200) throw Error(body.message);
    return body;
  };

  clearArticles = () => {
    this.setState(prevState => ({ articles: [] }))
  };

  stopPrinting = () => {
    this.shouldPrint['print'] = false
  }

  startPrinting = () => {
    this.shouldPrint['print'] = true;
    this.printArticles();
  }

  renderArticle(article) {
      return(
        <Article
          key={article.id}
          title={article.title}
          url={article.url}
          />
        )
  };

  render() {
    return(
      <div>
        <Controls
          clearArticles={this.clearArticles}
          stopPrinting={this.stopPrinting}
          startPrinting={this.startPrinting}
        />
        {this.state.articles && this.state.articles.map(article => {
          return this.renderArticle(article)
        })}
      </div>
    )
  };
};

export default ArticlesContainer;
