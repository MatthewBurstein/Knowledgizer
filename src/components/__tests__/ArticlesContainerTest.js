import React from 'react';
import ArticlesContainer from '../ArticlesContainer';
import { shallow, mount } from 'enzyme';

jest.useFakeTimers();

describe('ArticlesContainer', () => {
  const articles = [
    { id:1, title: 'some title', url: 'https://en.wikipedia.org/wiki/some title'},
    { id:2, title: 'another title', url: 'https://en.wikipedia.org/wiki/another title'},
    { id: 3, title: 'a third title', url: 'https://en.wikipedia.org/wiki/a third title' }
  ];
  const articlesContainer = shallow(<ArticlesContainer />);
  articlesContainer.setState({ articles });

  it('renders correctly', () => {
    expect(articlesContainer).toMatchSnapshot();
  });

  it('contains a Controls component', () => {
    const controls = articlesContainer.find('Controls');
    expect(controls).toHaveLength(1)
  })

  it('Contains an Article component for each article in `state`', () => {
    const articles = articlesContainer.find('Article');
    expect(articles).toHaveLength(3);
  });

  describe('.clearArticles()', () => {
    it('clears all articles from state', () => {
      articlesContainer.instance().clearArticles()
      expect(articlesContainer.state().articles).toHaveLength(0)
    });
  });

  describe('.stopPrinting', () => {
    it('sets state.shouldPrint to false', () => {
      articlesContainer.instance().stopPrinting();
      expect(articlesContainer.instance().shouldPrint.print).toBe(false)
    })
  })

  describe('delay and repeat article rendering functionality', () => {
    let mountedArticlesContainer, callApiSpy, didMountSpy, setStateSpy;
    beforeAll(() => {
      callApiSpy = jest.spyOn(ArticlesContainer.prototype, 'callApi').mockReturnValue(Promise.resolve({articles}));
      setStateSpy = jest.spyOn(ArticlesContainer.prototype, 'setState');
      mountedArticlesContainer = mount(<ArticlesContainer />);
    });

    it('calls call Api on mounting', () => {
      expect(callApiSpy).toHaveBeenCalledTimes(1);
    });

    it('calls setState once per second with the next article from this.articles', () => {
      expect(setStateSpy).not.toHaveBeenCalled();

      jest.runTimersToTime(1000);
      expect(setStateSpy).toHaveBeenCalledTimes(1);
      expect(mountedArticlesContainer.state().articles).toContain(articles[0]);

      jest.runTimersToTime(1000);
      expect(setStateSpy).toHaveBeenCalledTimes(2);
      expect(mountedArticlesContainer.state().articles).toContain(articles[1]);

      jest.runTimersToTime(1000);
      expect(setStateSpy).toHaveBeenCalledTimes(3);
      expect(mountedArticlesContainer.state().articles).toContain(articles[2]);
    });
  });

})
