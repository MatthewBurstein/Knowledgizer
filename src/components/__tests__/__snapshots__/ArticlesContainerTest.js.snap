// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticlesContainer renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ArticlesContainer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Article
          title="some title"
          url="https://en.wikipedia.org/wiki/some title"
        />,
        <Article
          title="another title"
          url="https://en.wikipedia.org/wiki/another title"
        />,
        <Article
          title="a third title"
          url="https://en.wikipedia.org/wiki/a third title"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "title": "some title",
          "url": "https://en.wikipedia.org/wiki/some title",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "class",
        "props": Object {
          "title": "another title",
          "url": "https://en.wikipedia.org/wiki/another title",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "3",
        "nodeType": "class",
        "props": Object {
          "title": "a third title",
          "url": "https://en.wikipedia.org/wiki/a third title",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Article
            title="some title"
            url="https://en.wikipedia.org/wiki/some title"
          />,
          <Article
            title="another title"
            url="https://en.wikipedia.org/wiki/another title"
          />,
          <Article
            title="a third title"
            url="https://en.wikipedia.org/wiki/a third title"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "title": "some title",
            "url": "https://en.wikipedia.org/wiki/some title",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "class",
          "props": Object {
            "title": "another title",
            "url": "https://en.wikipedia.org/wiki/another title",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "3",
          "nodeType": "class",
          "props": Object {
            "title": "a third title",
            "url": "https://en.wikipedia.org/wiki/a third title",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
